### PR TITLE
SANA Streaming Training Code

### PR DESCRIPTION
**Pull request by HeliosZhao**
Release both bidirectional and streaming distillation training code

### CODE REVIEW

File: configs/sana_streaming/train/sana_streaming_long_441_2b_720p.yaml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+beta2_critic: 0.999
+data_path: data/sana_streaming_long_489/manifest.jsonl
+batch_size: 1
+gradient_accumulation_steps: 1
+ema_weight: 0.99
+ema_start_step: 200
+log_iters: 100
+negative_prompt: "A chaotic sequence with misshapen, deformed limbs in heavy motion blur, sudden disappearance, jump cuts, jerky movements, rapid shot changes, frames out of sync, inconsistent character shapes, temporal artifacts, jitter, and ghosting effects, creating a disorienting visual experience."
+dfake_gen_update_ratio: 4
+image_or_video_shape: [1, 56, 128, 22, 40]
```
Comment (lawrence-cj) [inline configs/sana_streaming/train/sana_streaming_long_441_2b_720p.yaml:39]: This 489-stage config uses 56 latent frames. With the 8x temporal VAE stride, `(56 - 1) * 8 + 1 = 441`, so this trains on 441 rather than 489 source frames. Should this be 62?

File: tools/v2v/prepare_bidirectional_checkpoint.py [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+
+        q, k, v = tensor.chunk(3, dim=0)
+        if kind == "weight":
+            hidden_size = q.shape[0]
+            if q.shape != k.shape or q.shape != v.shape or q.shape[1] != hidden_size:
+                raise ValueError(
+                    f"{qkv_key} shape {tuple(tensor.shape)} does not describe three square Q/K/V projections."
+                )
+            # V2VBiGDNAttention represents Q and K as kernel-size-one Conv1d
+            # layers, while V remains a Linear layer.
```
Comment (lawrence-cj) [inline tools/v2v/prepare_bidirectional_checkpoint.py]: This converts every Q/K weight to Conv1d shape. The configured softmax layers use `nn.Linear`, so their Q/K weights must remain 2-D; the converted hybrid checkpoint may fail strict loading.

File: tools/v2v/prepare_bidirectional_checkpoint.py [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+"""Prepare a SANA-Video checkpoint for bidirectional V2V training.
```
Comment (lawrence-cj) [inline tools/v2v/prepare_bidirectional_checkpoint.py]: This conversion tool is not referenced by any config or documentation. Please document the supported workflow if it is public; otherwise it looks like one-off release code that should be removed.

File: diffusion/longsana/pipeline/sana_training_pipeline.py [hunk 4]
```diff
@@ -210,22 +312,40 @@ def generate_chunk_with_cache(
         exit_flags = self.generate_and_sync_list(num_chunks, steps, device)
 
         output = torch.zeros_like(latents)
-        previous_chunk_index = 0
-        for _kv_cache in self.kv_cache:
-            if _kv_cache[0][-1] is not None:
-                previous_chunk_index += 1
+        if self.v2v:
```
Comment (lawrence-cj) [inline diffusion/longsana/pipeline/sana_training_pipeline.py:313]: `requires_grad=False` only assigns an unused variable in this method; the exit-step generator call still builds an autograd graph. This may retain substantial memory during critic and warmup rollouts.

File: diffusion/longsana/trainer/longsana_trainer.py [hunk 5]
```diff
@@ -66,6 +68,38 @@ def start_new_sequence(self):
...
+                ]
+            except (OSError, RuntimeError, ValueError) as error:
+                rank = dist.get_rank() if dist.is_initialized() else 0
+                print(colored(f"[LongSANATrainer][rank {rank}] skipping source video: {error}", "red"))
+                skip_batch.fill_(1)
+
+            if dist.is_initialized():
+                dist.all_reduce(skip_batch, op=dist.ReduceOp.MAX)
+            if skip_batch.item():
+                return self.start_new_sequence()
```
Comment (lawrence-cj) [inline diffusion/longsana/trainer/longsana_trainer.py]: Recursively calling `start_new_sequence()` can recurse indefinitely on bad data and retain tensors from earlier stack frames. Please use a bounded iterative retry.

File: configs/sana_streaming/train/sana_streaming_bidirectional_2b_720p.yaml [hunk 6]
```diff
@@ -0,0 +1,108 @@
+task: v2v
+data:
+  data_dir:
+    example_data: data/sana_streaming_1k/data/example_data
```
Comment (lawrence-cj) [inline configs/sana_streaming/train/sana_streaming_bidirectional_2b_720p.yaml:4]: The default recipe points to this local dataset path, but the PR provides no dataset or download link/command. A clean checkout cannot run the documented bidirectional training.

File: diffusion/longsana/pipeline/sana_training_pipeline.py [hunk 7]
```diff
@@ -496,6 +626,152 @@ def inference_with_trajectory(
...
+        with the existing private API; cache tensors are allocated lazily by
+        the attention modules.
+        """
+        del batch_size, dtype, device
+        if num_chunks is None:
+            num_chunks = max(0, len(self.state["chunk_indices"]) - 1)
+        self.kv_cache = [self._empty_chunk_cache() for _ in range(num_chunks)]
+        return self.kv_cache
+
+    def _accumulate_v2v_kv_cache(self, kv_cache, chunk_idx):
```
Comment (lawrence-cj) [inline diffusion/longsana/pipeline/sana_training_pipeline.py]: This cache accumulation duplicates the implementation in `diffusion/scheduler/sana_streaming_sampler.py`. Keeping two copies may let training and inference cache behavior drift; please share one helper.

File: configs/sana_streaming/train/sana_streaming_bidirectional_2b_720p.yaml (hunk 6 above)
Comment (HeliosZhao) [reply configs/sana_streaming/train/sana_streaming_bidirectional_2b_720p.yaml:4]: Added the HF hyperlink in readme

File: configs/sana_streaming/train/sana_streaming_long_441_2b_720p.yaml (hunk 1 above)
Comment (HeliosZhao) [reply configs/sana_streaming/train/sana_streaming_long_441_2b_720p.yaml:39]: This stage should be 441 stage, 56 latent is correct. I revised every 489 to 441

File: diffusion/longsana/pipeline/sana_training_pipeline.py (hunk 4 above)
Comment (HeliosZhao) [reply diffusion/longsana/pipeline/sana_training_pipeline.py:313]: requires_grad is actually not used in this method. There is a torch.no_grad warp in the trainer when grad of this method is not used.

File: tools/v2v/prepare_bidirectional_checkpoint.py (hunk 3 above)
Comment (HeliosZhao) [reply tools/v2v/prepare_bidirectional_checkpoint.py]: Removed

File: diffusion/longsana/trainer/longsana_trainer.py (hunk 5 above)
Comment (HeliosZhao) [reply diffusion/longsana/trainer/longsana_trainer.py]: Add a 10-times max retry

File: tools/v2v/prepare_bidirectional_checkpoint.py (hunk 2 above)
Comment (HeliosZhao) [reply tools/v2v/prepare_bidirectional_checkpoint.py]: Removed

File: diffusion/longsana/pipeline/sana_training_pipeline.py (hunk 4 above)
Comment (lawrence-cj) [reply diffusion/longsana/pipeline/sana_training_pipeline.py:313]: The critic path is under `no_grad`, but the generator warmup at line 240 is not. `requires_grad=False` still returns a `CopySlices` grad_fn; please wrap that call or honor the flag.

File: diffusion/longsana/trainer/longsana_trainer.py (hunk 5 above)
Comment (lawrence-cj) [reply diffusion/longsana/trainer/longsana_trainer.py]: `skipped_batches_count` is never reset after a successful batch, so 11 failures across the whole run still abort training.

File: diffusion/longsana/pipeline/sana_training_pipeline.py (hunk 7 above)
Comment (lawrence-cj) [reply diffusion/longsana/pipeline/sana_training_pipeline.py]: Fixed in 5f0c48c0: training and inference now share `sana_streaming_cache.py`, with cache-window and promotion parity coverage.

File: diffusion/longsana/trainer/longsana_trainer.py (hunk 5 above)
Comment (lawrence-cj) [reply diffusion/longsana/trainer/longsana_trainer.py]: Fixed in c4dbe88d: retries now use a bounded local loop, reset after success, and no longer recurse. Added coverage for reset and 11 consecutive failures.

**Review by lawrence-cj — APPROVED** [review]

File: diffusion/longsana/model/streaming_sana_long.py [hunk 8]
```diff
@@ -318,6 +363,107 @@ def setup_sequence(
...
+
+        edited_chunk = rearrange(chunk.detach(), "b f c h w -> b c f h w")
+        gradient_mask = chunk_info.get("gradient_mask")
+
+        kwargs = {
+            "source_chunk_bcfhw": source_chunk,
+            "edited_chunk_bcfhw": edited_chunk,
+            "prompt_embeds": reverse_conditional_dict["prompt_embeds"],
+            "mask": reverse_conditional_dict.get("mask"),
+            "current_start_frame": chunk_info["chunk_start_frame"],
```
Comment (lawrence-cj) [inline diffusion/longsana/model/streaming_sana_long.py:410]: Keeping the two overlap frames makes reverse process `5,3,3` and writes those history frames into its cache again. If reverse should supervise only the 9 new frames, slice the overlap from both source and edited inputs; then `current_start_frame` remains correct.

File: diffusion/longsana/pipeline/sana_training_pipeline.py [hunk 9]
```diff
@@ -210,22 +310,40 @@ def generate_chunk_with_cache(
         exit_flags = self.generate_and_sync_list(num_chunks, steps, device)
 
         output = torch.zeros_like(latents)
-        previous_chunk_index = 0
-        for _kv_cache in self.kv_cache:
-            if _kv_cache[0][-1] is not None:
-                previous_chunk_index += 1
+        if self.v2v:
+            previous_chunk_index = current_start_frame // self.num_frame_per_block
```
Comment (HeliosZhao) [inline diffusion/longsana/pipeline/sana_training_pipeline.py:314]: Let's see the chunk is [0,5,8,11], for the third chunk, the current_start_frame is 8, previous_chunk_index is 8//3 = 2, which is correct.  previous_chunk_index is how many chunks exists before

File: diffusion/longsana/model/streaming_sana_long.py [hunk 10]
```diff
@@ -429,29 +576,48 @@ def generate_next_chunk(self, requires_grad: bool = True) -> Tuple[torch.Tensor,
...
+            image_vae_embeds=current_image_vae_embeds,
         )
 
         # Build the full chunk for loss computation
         if previous_frames is not None and overlap_frames_to_use > 0:
             # Concatenate specified overlap frames and newly generated frames
             full_chunk = torch.cat([previous_frames, generated_new_frames], dim=1)
+            previous_image_vae_embeds = self.state.get("previous_image_vae_embeds")
+            full_image_vae_embeds = (
+                torch.cat([previous_image_vae_embeds, current_image_vae_embeds], dim=2)
```
Comment (HeliosZhao) [inline diffusion/longsana/model/streaming_sana_long.py:613]: Fixed in [ca4099b](https://github.com/NVlabs/Sana/pull/421/commits/ca4099b2e2e94651a4e5518a46eec224e786b714)

**Review by lawrence-cj — APPROVED** [review]
The PR now looks great to me.
Thanks Yuyang!
